src/traits.rs

```diff
@@ -448,12 +444,34 @@ pub trait FindSubstring<T> {
 
 impl<'a,'b> FindSubstring<&'b [u8]> for &'a[u8] {
   fn find_substring(&self, substr: &'b[u8]) -> Option<usize> {
-    for (index,win) in self.windows(substr.len()).enumerate() {
-      if win == substr {
-        return Some(index)
+    let substr_len = substr.len();
+
+    if substr_len == 0 {
+      None
+    } else if substr_len == 1 {
+      memchr::memchr(substr[0], self)
+    } else {
+      let max = self.len() - substr_len;
+      let mut offset = 0;
+      let mut haystack = &self[..];
+
+      while let Some(position) = memchr::memchr(substr[0], haystack) {
```

Comment:
I think you miss the fact that the Memchr iterator doesn't iterate the occurrences but the position of these occurrences. Quoting the [documentation](https://docs.rs/memchr/1.0.1/memchr/struct.Memchr.html):
> Creates a new iterator that yields all positions of needle in haystack.

It also mentions that `type Item = usize`, and `find` returns `Option<Self::Item>` :).